### PR TITLE
Serialize deserialize admin boundaries

### PR DESCRIPTION
**Pull request by datanel**
needs https://github.com/georust/rust-geojson/pull/56 to be merged

TODO:
 - [x] integration tests

### CODE REVIEW

File: libs/mimir/src/objects.in.rs [hunk 1]
```diff
@@ -174,10 +174,48 @@ pub struct Admin {
...
+{
+    use geojson;
+    use serde::Deserialize;
+    use geojson::conversion::TryInto;
+
+    Option::<geojson::GeoJson>::deserialize(d).map(|option| {
+        option.and_then(|geojson| {
+            match geojson {
+                geojson::GeoJson::Geometry(geojson_geom) => {
+                    let geo_geom: geo::Geometry<f64> = geojson_geom.value.try_into().unwrap();
```
Comment (TeXitoi) [inline libs/mimir/src/objects.in.rs]: Can't this unwrap removed in favor of returning an error? Or matching `Ok(Multippolygon(m))` directly after?

**Comment by antoine-de** [comment]
the included Admin will be done in a separate PR

**Comment by antoine-de** [comment]
The admin refatoring is in fact in this PR

**Comment by antoine-de** [comment]
@TeXitoi we have pushed the modification

this PR still needs integartion tests

File: libs/bragi/Cargo.toml [hunk 2]
```diff
@@ -20,7 +20,8 @@ rs-es = "0.5"
 serde = "0.8"
 serde_json = "0.8"
 serde_derive = { version = "0.8", optional = true }
-geojson = { version="*", features=["with-serde"] }
+# we temporarily use our own git while gojson hasn't merged our PR on deserialization
```
Comment (TeXitoi) [inline libs/bragi/Cargo.toml]: typo

File: libs/bragi/src/query.rs [hunk 3]
```diff
@@ -52,21 +52,25 @@ pub fn make_place(doc_type: String, value: Option<Box<serde_json::Value>>) -> Op
...
                 serde_json::from_value::<mimir::Admin>(*v)
+                    .map_err(|err| warn!("Impossible to load ES result: {}", err))
                     .ok()
                     .and_then(|o| Some(mimir::Place::Admin(o)))
             }
             "poi" => {
                 serde_json::from_value::<mimir::Poi>(*v)
+                    .map_err(|err| warn!("Impossible to load ES result: {}", err))
                     .ok()
                     .and_then(|o| Some(mimir::Place::Poi(o)))
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]:
```rust
fn convert<T>(v: Json, f: fn(T) -> mimir::Place) -> Option<mimir::Place> {
    serde_json::from_value::<T>(*v)
        .map_err(|err| warn!("Impossible to load ES result: {}", err))
        .ok()
        .map(f)
}
match doc_type.as_ref() {
    "addr" => convert(*v, mimir::Place::Addr),
    "street" => convert(*v, mimir::Place::Street),
    "admin" => convert(*v, mimir::Place::Admin),
    "poi" => convert(*v, mimir::Place::Poi),
    _ => None,
}

File: src/admin_geofinder.rs [hunk 4]
```diff
@@ -34,16 +34,30 @@ use std::iter::FromIterator;
...
 use gst::rtree::{RTree, Rect};
 
+/// We want to strip the admin's boundary for the objects referencing it (for performance purpose)
+/// thus in the AdminGeoFinder we store an Admin without the boundary (the option is emptied)
+/// and we store the boundary aside
+struct AdminAndBoundary(Rc<Admin>, Option<geo::MultiPolygon<f64>>);
+
+impl AdminAndBoundary {
+    fn new(mut admin: Admin) -> AdminAndBoundary {
+        let b = admin.boundary.clone();
```
Comment (TeXitoi) [inline src/admin_geofinder.rs]:
```rust
let b = std::mem::replace(admin.boundary, None);
```

File: src/admin_geofinder.rs [hunk 5]
```diff
@@ -34,16 +34,30 @@ use std::iter::FromIterator;
 use std::rc::Rc;
 use gst::rtree::{RTree, Rect};
 
+/// We want to strip the admin's boundary for the objects referencing it (for performance purpose)
+/// thus in the AdminGeoFinder we store an Admin without the boundary (the option is emptied)
+/// and we store the boundary aside
+struct AdminAndBoundary(Rc<Admin>, Option<geo::MultiPolygon<f64>>);
```
Comment (TeXitoi) [inline src/admin_geofinder.rs]: I'd put that in the other sens, as (key, value)

File: libs/mimir/src/objects.in.rs [hunk 6]
```diff
@@ -204,10 +204,14 @@ fn custom_multi_polygon_deserialize<D>(d: &mut D)
...
                 geojson::GeoJson::Geometry(geojson_geom) => {
-                    let geo_geom: geo::Geometry<f64> = geojson_geom.value.try_into().unwrap();
+                    let geo_geom: Result<geo::Geometry<f64>, _> = geojson_geom.value.try_into();
                     match geo_geom {
-                        geo::Geometry::MultiPolygon(geo_multi_polygon) => Some(geo_multi_polygon),
-                        _ => None,
+                        Ok(geo::Geometry::MultiPolygon(geo_multi_polygon)) => Some(geo_multi_polygon),
+                        Ok(_) => None,
+                        Err(e) => {
+                            warn!("Error deserializing geometry: Message cool {}", e);
```
Comment (antoine-de) [inline libs/mimir/src/objects.in.rs]: we can do better than "message cool" :rofl:

File: tests/tests.rs [hunk 7]
```diff
@@ -183,12 +183,12 @@ fn all_tests() {
...
-    );
-    rubber_test::rubber_zero_downtime_test(ElasticSearchWrapper::new(&docker_wrapper));
-    rubber_test::rubber_custom_id(ElasticSearchWrapper::new(&docker_wrapper));
+    // bano2mimir_test::bano2mimir_sample_test(ElasticSearchWrapper::new(&docker_wrapper));
+    // osm2mimir_test::osm2mimir_sample_test(ElasticSearchWrapper::new(&docker_wrapper));
+    // osm2mimir_bano2mimir_test::osm2mimir_bano2mimir_test(
+    //     ElasticSearchWrapper::new(&docker_wrapper)
+    // );
+    // rubber_test::rubber_zero_downtime_test(ElasticSearchWrapper::new(&docker_wrapper));
+    // rubber_test::rubber_custom_id(ElasticSearchWrapper::new(&docker_wrapper));
```
Comment (TeXitoi) [inline tests/tests.rs]: ???

Comment (datanel) [inline tests/tests.rs]: bad rebase

**Pull request merged**